Correcciones

emanuelandrada opened the pull request:
Acá voy a agregar las correcciones al TP. No mergeen este pull request, lo haré yo cuando las correcciones estén hechas. Las correcciones háganlas en master (como vienen haciendo) y van a verse reflejadas aquí. También pueden responder con comentarios.

itManuel committed:
Create README.md

maller-glob committed:
Cambios en la indentación, y saqué el mensaje de agradecimiento fuera
del while.

maller-glob committed:
Desarme todo en funciones y puse qué tendría que hacer cada funcion.
Linda ensalada.

maller-glob committed:
Agregado un return a una funcion que declaraba bool

maller-glob committed:
Simplificando un poco

maller-glob committed:
Asignado 'borrar equipo' a IAN

ianalexis committed:
Una boludes

Agregue el 0-100 en defenza

ianalexis committed:
Cambio ventana y correccion de tildes.

maller-glob committed:
Agregué la definicion del TP en /doc, para que todos tengamos claro de qué hablamos.

maller-glob committed:
Funciones terminadas:
GuardarEquipo
VerEquipos

Funciones a revisar:
AgregarEquipo (nombres de equipo con espacios rompe todo)
YaEstaEquipo (por alguna causa, no encuentra cuando repite código)
MostrarEquipo (pega código y nombre como si fuese código al mostrar)

maller-glob committed:
algunas correcciones menores

maller-glob committed:
Corrección de VerEquipos

ianalexis committed:
Nombres y demas

caca

ianalexis committed:
Titulo

MateoTozzini committed:
Correción de Tildes

Commit:
Agrego comentario

Para ver como funciona esta mierda

Commit:
Vuelvo al original

maller-glob committed:
Abro archivo y cargo el array, al comenzar el programa.
Trunco el archivo y guardo el array, al finalizar el programa.

maller-glob committed:
Borre algunas funciones que ya no se usaban.

Commit:
Version de torneo de futbol sin array

ianalexis committed:
Inicio de eliminar

Esperando a lucas

ianalexis committed:
Revert "Inicio de eliminar"

This reverts commit 34ac2b783d8ad631c740247945f99eefd87ca468.

Commit:
Funcion modificar equipo funcionando

Commit:
Merge branch 'master' of https://github.com/itManuel/TorneoDeFutbol

Commit:
Agrego limite de 100 equipos

Commit:
Termino eliminar equipo

ianalexis committed:
Edite menu

Edite que agregar tambien diga editar

Commit:
FINAL

Commit:
Saco Equipo equipo innecesarios

ianalexis committed:
Correcciones tontas Version Final

Vfinal compilen y miren

ianalexis committed:
Final

maller-glob committed:
A ver cómo resolvemos ésto...

maller-glob committed:
O sea, va a fallar en cualquier cosa que no sea windows!!

maller-glob committed:
Y yo voy a terminar corrigiendo todo este, asi anda en todos lados.

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,156 @@
+	return opcion;
+}
+
+void ModificarEquipo(char id[4]){
```
¡Completar!

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,156 @@
+	// funcion Main.
+	// Manuel
+
+	/* abro el archivo al principio, cargo los equipos en un array y cierro el archivo. Trabajamos con los datos en memoria. */
```
Modularicen. El programa principal main podría quedarles así:

``` c++
in main ()
{
    Equipo equipos[MAX_EQUIPOS];
    int lineas = 0;
    CargarEquipos(equipos, lineas);
    MenuPrincipal(equipos, lineas);
    GuardarEquipos(equipos, lineas);
    return 0;
}
```

De esta manera es más legible qué hace cada parte.

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,156 @@
+		lineas++;
+	}
+	fclose(archivo);
+/* muestro lo que cargue del archivo */
```
No muestren el contenido del archivo al iniciar el programa.

itManuel replied:
je, quedó de cosas que fuimos haciendo para ver qué hacía... digamos que 'debug a los martillazos'

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,156 @@
+
+	FILE* archivo = fopen(FILENAME, "rb");
+	fread(&equipo[lineas],sizeof(equipo),1,archivo);
+	lineas++;
```
Están contando una línea de más, ya que si no pudo leer el `feof` retorna `false`, con lo cual no entra al ciclo pero contó la lectura. Quiten el incremento fuera del ciclo y siempre pongan la siguiente lectura al final del ciclo:

``` c++
fread(&equipo[lineas],sizeof(equipo),1,archivo);
while(!feof(archivo)){
    lineas++;
    fread(&equipo[lineas],sizeof(equipo),1,archivo);
}
```

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,156 @@
+				AgregarEquipo(lineas,equipo);
+				break;
+			case 2:
+				cout << "Esto hara que entres en el menu de eliminar equipo Y LO VA A HACER IAN" << endl;
```
Media pila Ian. :stuck_out_tongue:

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,156 @@
+	Equipo nuevoEquipo;
+	cout << "Ingrese el identificador (3 caracteres): ";
+	cin >> nuevoEquipo.id;
+	for(i=0;i<lineas;i++){
```
Subprograma. En vez de `if (agregar)` --> `if (EquipoExiste(equipos, lineas, nuevoEquipo.id))`.
O puede hacer algo como:

``` c++
int posEquipo = PosicionEquipo(equipos, lineas, nuevoEquipo.id);
if (posEquipo == lineas) { // equipo nuevo
    // agregar equipo nuevo
}
else { // equipo existente
    // preguntar si quiere modificar y modificarlo si dice que sí
}
```

De esta forma evitan buscar nuevamente dónde está, para modificarlo, si ya existe.

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,156 @@
+		}
+	}
+	if(agregar){
+		cout << "Ingrese el nombre (31 caracteres): ";
```
Subprograma. Así lo pueden reutilizar para la modificación de un equipo existente.

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,156 @@
+		cin >> nuevoEquipo.potenciaAtaque;
+		cout << "Ingrese potencia de defensa (0-100): ";
+		cin >> nuevoEquipo.potenciaDefensa;
+		lineas++;
```
Tienen que sumar 1 después de agregar al arreglo, si no lo agregan en una posición incorrecta.

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,156 @@
+	if(agregar){
+		cout << "Ingrese el nombre (31 caracteres): ";
+/* por alguna razon, esta linea de abajo solo toma la primer palabra, hasta el espacio. */
+		cin >> nuevoEquipo.nombre;
```
Este tipo de ingreso toma solo una palabra. Para leer una línea completa pueden usar algo como lo siguiente:

``` c++
string nombre; // variable auxiliar para leer una línea completa
getline(cin, nombre); // lee una línea vacía que quedó pendiente del último ingreso.
getline(cin, nombre); // lee una línea desde el teclado
strncpy(nuevoEquipo.nombre, nombre.c_str(), sizeof(nuevoEquipo.nombre)); // copia los caracteres al registro
nuevoEquipo.nombre[sizeof(nuevoEquipo.nombre) - 1] = '\0'; // fuerza a que tenga a lo sumo 31 caracteres
```

emanuelandrada commented:
Entiendo que TorneoDeFutbol2.cpp no lo están usando, ¿no?

itManuel commented:
Fué un intento de utilizar funciones sólo windows para hacer cosas, y con
ese intento se logró completar todas las funciones.

Ahora, volviendo al TorneoDeFutbol.cpp, estamos teniendo problemas con:
    for(i=0;i<lineas;i++){
        cout << "Guardando: " << equipos[i].id << "|" << equipos[i].nombre
<< "|" << equipos[i].potenciaAtaque << "|" << equipos[i].potenciaDefensa <<
endl;
        fwrite(&equipos[i], sizeof(equipos), 1, archivo);
    }

ya que como es esperable, imprime por pantalla correctamente todos los
datos, pero al volver a leer el archivo sólo guardó los id, por lo que
puedo suponer que en realidad sólo eso es lo que está guardando.

Alguna idea?

Gracias,

2015-06-14 0:44 GMT-03:00 Emanuel notifications@github.com:

> Entiendo que TorneoDeFutbol2.cpp no lo están usando, ¿no?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/itManuel/TorneoDeFutbol/pull/1#issuecomment-111777915
> .

## 

Manuel

maller-glob committed:
modularizando y corrigiendo cosas de los contadores. Limpieza general y particular del MAIN().

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,162 @@
+
+	for(i=0;i<lineas;i++){
+		cout << "Guardando: " << equipos[i].id << "|" << equipos[i].nombre << "|" << equipos[i].potenciaAtaque << "|" << equipos[i].potenciaDefensa << endl;
+		fwrite(&equipos[i], sizeof(equipos), 1, archivo);
```
El problema que tienen acá es el `sizeof`. `sizeof` devuelve el espacio que ocupa en memoria el argumento. En este caso pasaron `equipos`, que es un arreglo sin tamaño. Los arreglos sin tamaño (solo como parámetros pueden no tener tamaño) son en realidad punteros al primer elemento del arreglo, por lo tanto acá el `sizeof` devuelve el tamaño del puntero. Si están en una máquina de 32 bits, esto es 4 bytes, que coincide con los cuatro bytes del identificador.
Lo corrigen poniendo `sizeof(Equipo)` (toma el tamaño del `struct Equipo`) o `sizeof(equipos[i])` (toma el tamaño del elemento del arreglo, como los elementos son de `struct Equipo` devuelve ese tamaño).

emanuelandrada commented on src/TorneoDeFutbol.cpp:
```diff
@@ -1,56 +1,162 @@
+	fread(&equipos[lineas],sizeof(equipos),1,archivo);
+	while(!feof(archivo)){
+		lineas++;
+		fread(&equipos[lineas],sizeof(equipos),1,archivo);
```
Acá les pasa lo mismo que al guardar. Corrijan el `sizeof`.

maller-glob committed:
Corrección en fread y fwrite

maller-glob committed:
todas las funcionalidades terminadas

Commit:
Agregue Menu Primario y Secundario

Commit:
Agregue simulacion de partidos y menus

Agregue Menus generales, generacion de partidos, simulacion de partidos
y Ver partidos por fecha. Hay que mejorar algunas cosas pero la base
esta.

emanuelandrada commented:
Las correcciones del TP 1 están bien.
En el TP 2 usen alguna estructura dinámica.

Commit:
Cambie todo a listas

maller-glob committed:
Algunas correcciones de logica

maller-glob committed:
borrando cosas que no tienen que estar (un exe??)

maller-glob committed:
algunos cambios para hacer que funcione bien los partidos y las simulaciones

Commit:
Algunas Boludeces

emanuelandrada commented:
Toda la parte de listas está igual a otro grupo que entregó unos días antes. Tienen los mismos errores que la primer entrega que hicieron. ¿Me pueden explicar qué pasó?
